### PR TITLE
application gateway waf config

### PR DESCRIPTION
**Pull request by azhao155**
# Conflicts:
#	tools/AzureRM/AzureRM.psm1

## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log)
- [x] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr)
    - [x] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md)

### CODE REVIEW

**Comment by azhao155** [comment]
@MikhailTryakhov Could you please generate the latest sdk, the validation will failed with the local network nuget package?

**Comment by MikhailTryakhov** [comment]
@azhao155 please review [process](https://microsoft.sharepoint.com/teams/WAG/AzureNetworking/_layouts/OneNote.aspx?id=%2Fteams%2FWAG%2FAzureNetworking%2FOneNotes%2FNRP%20migrated&wd=target%28Hydra%20%2B%20SDK.one%7C96F7B95B-A378-4CA5-903B-C88F894873EF%2FSwagger%20-%3E%20SDK%20-%3E%20PowerShell%20Process%20Overview%7C71B816ED-DAE5-4907-94D8-DE7EE13278A0%2F%29
onenote:https://microsoft.sharepoint.com/teams/WAG/AzureNetworking/OneNotes/NRP%20migrated/Hydra%20+%20SDK.one#Swagger%20-%3e%20SDK%20-%3e%20PowerShell%20Process%20Overview&section-id={96F7B95B-A378-4CA5-903B-C88F894873EF}&page-id={71B816ED-DAE5-4907-94D8-DE7EE13278A0}&end) how to work with monthly branches

**Comment by azhao155** [comment]
@MikhailTryakhov i only update the .csproj for network, not sure why other resources will fail.. Could you give me some help?

**Comment by EvgenyAgafonchikov** [comment]
Hi @azhao155, I've had a look into failures on CI, please find details below:
```
Failed   Microsoft.Azure.Commands.Compute.Test.ScenarioTests.ContainerServiceTests.TestContainerServiceUpdate
Error Message:
 System.Management.Automation.RuntimeException : Test failed due to a non-empty error stream. First error: PowerShell Error Record: Could not load file or assembly 'Microsoft.Azure.Management.Network, Version=19.6.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'.
```

Compute test project and several others (Sql, Dns, RecoveryServices.Backup, KeyVault) also depend on Networking SDK. So the version should be updated for them too.
I would send you email with the link to the tools that simplify updating such refs.

```
Failed   Microsoft.Azure.Commands.Compute.Test.ScenarioTests.VirtualMachineScaleSetTests.TestVirtualMachineScaleSetForceUDWalk
Error Message:
 System.Management.Automation.CommandNotFoundException : The term 'New-AzVirtualNetworkSubnetConfig' is not recognized as the name of a cmdlet, function, script file, or operable program.
```

It looks like mapping between old syntax (\*-AzureRM\*) and new one (\*-Az\*) is missing. I see it was added as a build step, so no manual action needed now. @maddieclayton, could you please have a look what could cause the failure above? I see related mapping under [this file](https://github.com/Azure/azure-powershell/commit/fe404fe0754390ca8abe1e648fbf730339b6a05a#diff-74f9c4e35c15c8bdebb85398606cf037R1435)

One more thing that could be found in [build artifacts](https://azuresdkci.westus2.cloudapp.azure.com/job/powershell/8098/artifact/src/Package/HelpIssues.csv/*view*/) is
```
"Assembly","HelpFile","Target","Severity","ProblemId","Description","Remediation"
"","","Microsoft.Azure.Commands.Network.NewAzureApplicationGatewayFirewallExclusionConfigCommand","1","6050","Help missing for cmdlet New-AzureRmApplicationGatewayFirewallExclusionConfig implemented by class Microsoft.Azure.Commands.Network.NewAzureApplicationGatewayFirewallExclusionConfigCommand","Add Help record for cmdlet New-AzureRmApplicationGatewayFirewallExclusionConfig to help file."
```
Please follow the steps from [there](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md) to re-generate help files.

**Comment by maddieclayton** [comment]
@azhao155 @EvgenyAgafonchikov To address the questions aimed at me, it looks like the new cmdlet is not included in the psd1 files for network, which is why the test can't find the cmdlet.

**Comment by azhao155** [comment]
@EvgenyAgafonchikov i tried add 19.6.0 reference in the compute/... csproj, but it still failed, so i revert back.

**Comment by EvgenyAgafonchikov** [comment]
@azhao155, I've opened PR to you fork that fixes significant part of the issues: https://github.com/azhao155/azure-powershell/pull/1, feel free to merge.

What was done:
- Version update in required projects (this fixed unrelated Compute tests failures, next items are fixes for other failures. Look for "fail" in this [job](https://travis-ci.org/Azure/azure-powershell/builds/436590787?utm_source=github_status&utm_medium=notification))
- Generating help file for the cmdlet, adding ref to the table of contents
- Fixing cmdets signature issues (requirement is to name parameters in singular form)
- Adding Signature exclusion for ShouldProcess
- [Updating psd1 file for "new-syntax" cmdlets](https://github.com/Azure/azure-powershell/pull/7455/files#diff-013e5e290ccacfd6af79225b5c6de2d3)

There are still things that need to be done after merging it:

1. Please lines containing "Placeholder" in help files with real descriptions for the feature added. E.g. [this](https://github.com/azhao155/azure-powershell/pull/1/files#diff-78a663ce09f9d30f8b790b9a3fbc3b64R441), [this one and other in this md file](https://github.com/azhao155/azure-powershell/pull/1/files#diff-e6af755cead9d69de66acf7afc2d9457R11)

2. The only type failure I see in tests currently is
```
Failed   Commands.Network.Test.ScenarioTests.NetworkProfileTests.TestContainerNetworkInterfaceConfigCRUD
Error Message:
 System.Management.Automation.ActionPreferenceStopException : The running command stopped because the preference variable "ErrorActionPreference" or common parameter is set to Stop: Operation returned an invalid status code 'Accepted'
```
Likely you need to pull the latest changes in rest-api-specs repo and regenerate your LocalFeed SDK. I believe this is the [commit that should fix this](https://github.com/Azure/azure-rest-api-specs/commit/4d1b5ac80425123cd271697fe46df2a0233ac3e6#diff-24aa859bca03c0402c202d43acee6d1b)

3. Possible something else left. Tracking this on my [temporary PR](https://github.com/Azure/azure-powershell/pull/7455) that contains your updates and my fixes both.

**Comment by azhao155** [comment]
@EvgenyAgafonchikov all check passed. Thanks a lot for the help. what will be the next step before merge?

**Comment by EvgenyAgafonchikov** [comment]
@azhao155, I believe the next step is review from maintainers side. You could update tag to "needs-review" to make the PR more visible.

**Comment by azhao155** [comment]
@MikhailTryakhov Could you please take a look since merge validation is all passing right now

**Comment by asridharan** [comment]
@MikhailTryakhov pinging again on this? We  want to get this into the Oct 20 release cycle so wanted to some feedback on this in case we need to revise.

**Comment by MikhailTryakhov** [comment]
@azhao155  @asridharan  I'm not the right  person to review, PS team should

**Comment by asridharan** [comment]
Thanks @MikhailTryakhov .

**Comment by praries880** [comment]
design review : https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/175

**Comment by MikhailTryakhov** [comment]
@azhao155  please resolve merge conflicts

**Comment by azhao155** [comment]
@praries880 have updated PR based on the review, Please check

File: src/ResourceManager/Network/Commands.Network/ApplicationGateway/FirewallConfiguration/AzureApplicationGatewayWebApplicationFirewallConfigurationBase.cs [hunk 1]
```diff
@@ -49,8 +49,33 @@ public class AzureApplicationGatewayWebApplicationFirewallConfigurationBase : Ne
         [Parameter(
                Mandatory = false,
                HelpMessage = "The disabled rule groups.")]
+        [Alias(DisabledRuleGroupsAlias)]
         [ValidateNotNullOrEmpty]
-        public List<PSApplicationGatewayFirewallDisabledRuleGroup> DisabledRuleGroups { get; set; }
+        public List<PSApplicationGatewayFirewallDisabledRuleGroup> DisabledRuleGroup { get; set; }
+
+        [Parameter(
```
Comment (praries880) [inline src/ResourceManager/Network/Commands.Network/ApplicationGateway/FirewallConfiguration/AzureApplicationGatewayWebApplicationFirewallConfigurationBase.cs]: you need to generate the help files for the cmdlets that this base class is used in as well.. as we added a  few params here.

**Comment by azhao155** [comment]
@praries880 Please review the change

**Comment by asridharan** [comment]
@praries880 @MikhailTryakhov  the SDK was committed to `network-october-release` branch:
https://github.com/Azure/azure-rest-api-specs/commit/69fc43dc5c40cec6f09b1ee7b895641dee74aa99#diff-fd8b5412b2cafcbab576a5c0ca24f6b2
So would a nuget package for the SDK automatically be generated and after that we remove the private SDK feed?

**Comment by praries880** [comment]
@azhao155  instructions on how to release an SDK are [here](https://github.com/Azure/azure-sdk-for-net/blob/psSdkJson6/Documentation/code-generation-instructions.md#how-to-signpublish-bits)

@MikhailTryakhov can the SDK be published?

**Pull request merged**